src/Proto.Cluster/ClusterExtension.cs

```diff
@@ -31,22 +31,33 @@ public static Task<T> ClusterRequestAsync<T>(this IContext context, string ident
         {
             var cluster = context.System.Extensions.Get<Cluster>();
             //call cluster RequestAsync using actor context
-            return cluster.RequestAsync<T>(identity, kind, message, context, ct);
+            return cluster!.RequestAsync<T>(identity, kind, message, context, ct);
         }
 
-        public static Props WithClusterInit(this Props props, Cluster cluster, ClusterIdentity clusterIdentity, ActivatedClusterKind activatedClusterKind)
+        public static Props WithClusterInit(
+            this Props props,
+            Cluster cluster,
+            ClusterIdentity clusterIdentity,
+            ActivatedClusterKind activatedClusterKind
+        )
         {
-            return props.WithReceiverMiddleware(
-                baseReceive =>
-                    (ctx, env) => {
-                        return env.Message switch
-                        {
-                            Started => HandleStart(baseReceive, ctx, env),
-                            Stopped => HandleStopped(baseReceive, ctx, env),
-                            _       => baseReceive(ctx, env)
-                        };
+            return props
+                .WithContextDecorator(context => {
+                        context.SetMetadata(clusterIdentity);
```

Comment:
This is where the recursion happens?

Reply:
Yes, since setting metadata requires extras to be instantiated